causync.py

```diff
@@ -0,0 +1,291 @@
+            dst = os.path.abspath("{dst}/{bd}_inc_{d}".format(dst=self.dst,
+                                                              bd=inc_basedir_dirname,
+                                                              d=curdate))
+
+        if self.incremental and self.incremental_basedir:
+            extra_flags = "--link-dest={}".format(self.incremental_basedir)
+
+        cmd = "rsync {f} {ef} {src} {dst}".format(f=self.config.RSYNC_FLAGS,
+                                                  ef=extra_flags,
+                                                  src=self.src_abs,
+                                                  dst=dst)
+
+        self.logger.info("syncing {} to {}".format(self.src_abs, dst))
+        self.create_lockfile()
+
+        result = subprocess.check_output(cmd, shell=True).decode()
+        self.logger.debug(result)
+
+        self.remove_lockfile()
+        self.logger.info("sync finished")
+
+        return result
+
+    def run_cleanup(self, incremental=False):
+        """ This function is executed when the task argument is 'cleanup'. """
+
+        dirnames = os.listdir(self.dst)
+
+        self.logger.debug(dirnames)
+
+        dirdates = []
+
+        if incremental:
+            basename = self.incremental_basedir.rsplit('/', 1)[1]
+        else:
+            basename = self.src_abs.split('_', 1)[0]
+
+        self.logger.debug(basename)
+
+        for dirname in dirnames:
+            if basename not in dirname:
+                continue
+            elif incremental and "_inc_" not in dirname:
+                continue
+            elif incremental and basename in dirname and "_inc_" in dirname:
+                try:
+                    dirdate = "_".join(dirname.rsplit('_', 2)[1:])
+                    dirdate = datetime.strptime(dirdate, config.DATE_FORMAT)
+                    dirdates.append(dirdate)
+                except ValueError as e:
+                    self.logger.debug(e)
+            else:
+                try:
+                    dirdate = "_".join(dirname.rsplit('_', 2)[1:])
+                    dirdate = datetime.strptime(dirdate, config.DATE_FORMAT)
+                    dirdates.append(dirdate)
+
+                except ValueError as e:
+                    self.logger.debug(e)
+
+        dirdates.sort()
+
+        keep = dirdates[-5:]
+        delete = dirdates[:-5]
+        self.logger.debug("keep={}".format(keep))
+        self.logger.debug("delete={}".format(delete))
+
+        if incremental:
+            rmstring = "{dst}/{bn}_inc_{d}"
+        else:
+            rmstring = "{dst}/{bn}_{d}"
+
+        for dirname in delete:
+            rmtree(rmstring.format(dst=self.dst.rstrip('/'),
+                                   bn=basename,
+                                   d=dirname.strftime(config.DATE_FORMAT)))
+
+        self.logger.info("successfully deleted old backups")
+
+    def create_lockfile(self):
```

Comment:
@gabalz, our initial specification mentions a lock directory, but this creates a lock file instead. It feels safer and easier to find (for example with `find . -type f`). Is this a good/better solution, or should I change it to a directory?

Reply:
Hmm, I don't see any problem with a lock file, please keep it for now.